doc/specs/Authentication.md

```diff
@@ -0,0 +1,477 @@
+# Authentication system
+
+## Metadata
+
+Author: DutchDude
+
+### Changelog
+
+**2015-07-11**
+
+* Incorporated feedback from GitHub after public release
+
+**2014-12-31**
+
+* Added second half of use cases
+* Added a bunch of business rules
+
+**2014-12-24**
+
+* Initial draft with first half
+
+## Table of contents
+
+1. Motivation
+1. Business rules
+1. Use cases
+1. Data model
+
+## Motivation
+
+The site is a walled garden, in which its community shares lots of great content.
+To keep the community in great shape and to ensure quality control, the website
+is not publicly accessible. Restricted access ensures that bad people are (mostly)
+kept out and keeps the site community and economy healthy.
+
+## Business rules
+
+Business rules are principles that should be honored throughout the application.
+They are not technical requirements, but specify what is needed to "help the
+organization achieve its goals." (Wikipedia)
+
+* Users need always be identifiable
+* For each user, it should be recorded when they last accessed the site.
+* The devices from which a user logs in should be recorded. This information
+  should include user agent, source IP address, authentication method and time.
+* If a user transfers a session to a different IP address, the new information
+  should also be recorded. (cookie stealing, proxies, 3g connections, etc)
+* Accessing the website from certain addresses may be prohibited by staff.
+* Authentication should always happen through a secure channel.
+* Authentication tokens such as passwords should be stored in a non-reversible
+  manner.
+* Users should be IP-blocked for an amount of time if they fail authentication
+  or try to reset a password  more than a configurable number of times.
+* Users should be informed of the ways they can reset passwords or bypass
+  secondary authentication, even if they have not failed authentication yet.
+* Resets of passwords should expire all active sessions and
+  authentication tokens for that user.
+* Password reset links should expire after a set number of hours.
+* Users may set up a secondary means of authenticating. If they have set
+  this up, they must be asked to further authenticate once they've authenticated
+  using their username and password.
+* If users set up secondary authentication they should be provided with
+  a set of one-time passwords to bypass secondary authentication.
+* If a user loses their secondary authentication method and loses their one-time
+  passwords, they will need to contact moderators through IRC to disable secondary
+  authentication.
+* Users may have at most one secondary means of authentication.
```

Comment:
These requirements were just copied of what's currently asked of passwords. I'm open to better suggestions.

Putting on my cryptographer-hat[1], I'm not really a big fan of that XKCD, however. Everything hangs on people choosing *random* words. People suck at that and then I'm not sure it's much better than these requirements.

[1]: I'm no Schneier, but I've got some experience with this stuff.